src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -865,65 +865,73 @@ private int ReadBuffer(Span<char> userBuffer, out bool readToUserBuffer)
             return task;
         }
 
-        private async Task<string?> ReadLineAsyncInternal()
+        private Task<string?> ReadLineAsyncInternal()
         {
-            if (_charPos == _charLen && (await ReadBufferAsync(CancellationToken.None).ConfigureAwait(false)) == 0)
-            {
-                return null;
-            }
-
-            StringBuilder? sb = null;
+            return _charPos == _charLen
+                ? NoDataInTheBuffer()
+                : ConsumeBufferedData();
 
-            do
+            async Task<string?> NoDataInTheBuffer()
             {
-                char[] tmpCharBuffer = _charBuffer;
-                int tmpCharLen = _charLen;
-                int tmpCharPos = _charPos;
-                int i = tmpCharPos;
+                if (await ReadBufferAsync(CancellationToken.None).ConfigureAwait(false) == 0)
+                    return null;
 
+                return await ConsumeBufferedData().ConfigureAwait(false);
+            }
```

Comment:
Doesn't this make the NoDataInTheBuffer case more expensive?  Is that just not common?

Reply:
It does but it might not be the most used one case, at least that's what I hope for. The benchmarks -> dotnet/performance@529f33c#diff-64fc887dbc1598f2ea871592946f4077593925f80b22cab6f586ea7e1e1b0723